### DISCUSSION TITLE
Dynamic Routing fail when build

### DISCUSSION THREAD

**Discussion by mankinchi** [vercel/next.js#18942] [Help] [answered]
Hi, 

I'm having a problem with building nextJS with dynamic routing. It keeps failing and I don't quite understand why. This is the single page I have.

```
import Axios from 'axios';

const Store = (props) => {
	return (
		<>
			<p>Store name is {props.name}</p>
			<p>Description is {props.description}</p>
			<img src={`${process.env.cmsAPIRoot}${props.logo.small.url}`}/>
		</>
	)
}

export async function getStaticPaths() {
	const { data } = await Axios.get(`${process.env.cmsAPIRoot}/stores`);

	const paths = data.map((store) => ({
		params: {
			id: store.id.toString(),
		},
	}));

	return {
		paths,
		fallback: true,
	}
}

export async function getStaticProps(context) {
	const { params: { id }} = context;

	const { data } = await Axios.get(`${process.env.cmsAPIRoot}/stores/${id}`);

	console.log(id);

	const {
		name,
		description,
		logo: {
			formats: logo,
		},
	} = data;

	return {
		props: {
			name,
			description,
			logo,
		},
		revalidate: 1,
	}
}

export default Store;
```

I have checked the paths to ensure that there are no weird path in there
```
[
  { params: { id: '1' } },
  { params: { id: '2' } },
  { params: { id: '3' } },
  { params: { id: '4' } }
]
```

And this is the error
```
Generating static pages (0/6)
Error occurred prerendering page "/stores/[id]". Read more: https://err.sh/next.js/prerender-error
TypeError: Cannot read property 'small' of undefined
```

If I switched `props.logo.small.url` to `props.logo?.small.url` then it will run smoothly. I have `console.log` to see what's the issue and there seems to be a problem during build (happen only 1 time) that page props are completely empty (`{}`), which leads it to fail. How can I overcome the issue? I have expected the page to wait for the `getStaticProps` to complete before render.

**Comment by jamesmosier** [comment, marked as answer]
Hi @mankinchi. You'll want to implement something similar to the [docs](https://nextjs.org/docs/basic-features/data-fetching#fallback-pages) for `isFallback`. Then your page will show a loading state while it is waiting to fetch the data (for non-static generated pages at build time).

```js
const Store = (props) => {
  const router = useRouter()

  // If the page is not yet generated, this will be displayed
  // initially until getStaticProps() finishes running
  if (router.isFallback) {
    return <div>Loading...</div>
  }

  // Render content...
}
```

**Reply by mankinchi** [reply]
why is it doing this? Is it to generating the fallback page? I.e `[id]` page

**Reply by jamesmosier** [reply]
If you are fetching all `stores` in `getStaticPaths` then you can remove `fallback: true` altogether. [fallback](https://nextjs.org/docs/basic-features/data-fetching#the-fallback-key-required) is only needed when you have paths that are not statically generated at build time.

**Comment by mankinchi** [comment]
I’m building a post page and therefore, new page will be rendered during runtime. I think I understand it correctly, the problem is causing by the fallback behaviour, which is generating a page with empty props

Kind regards,

Tri Nguyen
Software Developer

Kepler Analytics | Tools Built for Retail Success

m: +61 449 258 526 ( tel:+61%20449%20258%20526 )
w: kepleranalytics.com ( https://share.polymail.io/v1/z/b/NWZhOGFhZTZmYjgw/HMdJ1mBXVK_tuuYtIgOissjbMoXCfa1ewBGykQ0YA3NX8Zhp3uklhP8WmCfxQ7EiEwthCmbZXyjorfVotVQDnW2GB4LskbKttYqgumce0u4pXgzZal3KJwq25u4XpQIVyDBds3iAVVRrbyCBHDHJjShG3QsciqH_Z5ACkuQD_MvV4Z1srsuAOWONOmwl-G-oE1Do2A84wRdQt4Czo1Gjr-KlSfJzibM0qJ8PlzJNwEshk1n4HhhtmA== )
a: 57 Victoria Street, Fitzroy, VIC 3065, Australia
